### PR TITLE
Implement special parameter to define Key Usage

### LINKED ISSUE #285: Key generation should allow specifying key usage
When generating a new key, the application should be able to specify key usage in the context parameters (sign/verify, encrypt/decrypt). Current all keys are generated with all usage allowed.

But it is not recommended to generate RSA private keys that can be used both for signing and decryption.

### PR DESCRIPTION
**Pull request by simo5**
This allows an application finer control on what usage will be allowed for the key.

Fixes #285

### CODE REVIEW

**Comment by simo5** [comment]
@fabled will this work for you?

**Comment by simo5** [comment]
Tests are missing switching to draft

**Review by fabled — COMMENTED** [review]
Looks mostly good. The `toklen` related issue needs fixing.

File: src/keymgmt.c [hunk 1]
```diff
@@ -334,6 +343,118 @@ static CK_RV common_gen_callback(void *cbarg)
...
+            sig = true;
+        } else if (strncmp(str, "keyAgreement", toklen) == 0) {
+            der = true;
+        } else if (strncmp(str, "keyEncipherment", toklen) == 0) {
+            wrap = true;
+        } else {
+            return CKR_ARGUMENTS_BAD;
+        }
+        str = p;
+        len -= toklen;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/keymgmt.c:449]: good catch, although this means when there is a single token we'd get a negative len ... but in that case str will be NULL so we won't use it anymore, so perhaps that's fine.

File: src/keymgmt.c [hunk 2]
```diff
@@ -334,6 +343,118 @@ static CK_RV common_gen_callback(void *cbarg)
...
+    for (int i = 0; i < tsize; i++) {
+        switch (template[i].type) {
+        case CKA_ENCRYPT:
+        case CKA_DECRYPT:
+            if (enc) {
+                template[i].pValue = DISCARD_CONST(&val_true);
+            } else {
+                template[i].pValue = DISCARD_CONST(&val_false);
+            }
+            break;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/keymgmt.c:368]: this is a good point, I'll take a look.

**Comment by simo5** [comment]
@fabled updated the relevant code you pointed out, PTAL

**Review by sahanaprasad07 — COMMENTED** [review]
Looks good, but I'll wait for tests.

**Comment by simo5** [comment]
Updated with tests

**Comment by fabled** [comment]
Looks good to me. Thank you!

**Review by fabled — APPROVED** [review]

**Pull request merged**